### PR TITLE
added twis add/list/clear functionality for this week in servo action…

### PR DESCRIPTION
**Pull request by avadacatavra**
Did some testing in #avadacatavra

Will allow you to have crowbot store your accomplishments for this week in servo (issue #49)

r+ @jdm

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -191,6 +191,62 @@ var handlerWrapper = module.exports.handlerWrapper = function handlerWrapper(pin
       return;
     }
 
+    //This Week in Servo
+    if (message.indexOf('twis ') > -1) {
+      try {
```
Comment (jdm) [inline server.js]: I'm not super keen on this all-encompassing try/catch. The error logs already will get a message for uncaught exceptions, so we don't really gain anything here.

File: server.js [hunk 2]
```diff
@@ -191,6 +191,62 @@ var handlerWrapper = module.exports.handlerWrapper = function handlerWrapper(pin
...
+        var command = original_message.match(/(twis)(.*)/i)[2].trim().match(/([^ ]*)/);
```
Comment (jdm) [inline server.js]: What about this instead? I find it easier to read than the regexes.
```js
var command = original_message.split(' ').splice(1);
var action = command[0].toLowerCase();
var message = command.splice(1).join(' ');
```

File: server.js [hunk 3]
```diff
@@ -191,6 +191,62 @@ var handlerWrapper = module.exports.handlerWrapper = function handlerWrapper(pin
...
+
+                var verbs = [" accomplished ",
+                             " concluded ",
+                             " finished ",
+                             " managed ",
+                             " performed ",
+                             " did ",
+                             " won ",
+                             " realized ",
+                             " produced "];
```
Comment (jdm) [inline server.js]: I'd probably leave the spaces out and add them to the final string instead.

**Comment by jdm** [comment]
Sorry for sitting on this!

**Comment by avadacatavra** [comment]
@jdm made the changes

**Comment by jdm** [comment]
Let's add a test as well. That would make me more confident that my suggestions for rewriting the string parsing actually do what I expect.

File: server.js [hunk 4]
```diff
@@ -191,6 +191,53 @@ var handlerWrapper = module.exports.handlerWrapper = function handlerWrapper(pin
...
+
+              var verbs = ["accomplished ",
+                           "concluded ",
+                           "finished ",
+                           "managed ",
+                           "performed ",
+                           "did ",
+                           "won ",
+                           "realized ",
+                           "produced "];
```
Comment (jdm) [inline server.js]: We can remove the trailing spaces as well and add it to the final string.

File: server.js [hunk 5]
```diff
@@ -191,6 +191,53 @@ var handlerWrapper = module.exports.handlerWrapper = function handlerWrapper(pin
...
+          twisForUser.push({"from": from, "message": message});
+
+          twisStorage.setItemSync(from, twisForUser);
+          bot.say(to, "Done and done.");
+        } else {
+          bot.say(to, "Sorry. I can't add this. Did you include what you did?");
+        }  
+      } else if (action.indexOf('list') > -1) {       //TODO maybe make it all in one message?
+          bot.say(to, "This Week in Servo!");
+          var twisForUser = twisStorage.getItemSync(from);
```
Comment (jdm) [inline server.js]: Oh, does this mean that there's a separate TWiS list per user that added items? Like, if both ms2ger and I use `twis add someone did something`, if I use `twis list` I'll only see the entries that I added? If so, we should change that to have a single global store instead.

File: server.js [hunk 6]
```diff
@@ -402,6 +449,27 @@ var pingStorage = {
   removeItemSync: storage.removeItemSync
 }
 
+storage.initSync({
```
Comment (jdm) [inline server.js]: This still isn't using the factory constructor...

Comment (jdm) [inline server.js]: What the heck, I _swear_ I left a comment about using https://github.com/simonlast/node-persist#createoptions---synchronous-static-method but there's no record of it on this page.

File: server.js (hunk 5 above)
Comment (avadacatavra) [inline server.js]: This is the output:
15:30 <crowbot1> This Week in Servo!
15:30 <crowbot1> avadacatavra performed fixed crowbot
15:30 <crowbot1> avadacatavra_test finished changed names

It tracks who did what and stores it by name, then outputs all.

Comment (jdm) [inline server.js]: I see. I think it's important to support storing snippets about people other than the user interacting with crowbot, too.

**Comment by KiChjang** [comment]
What's going on with this?

**Comment by avadacatavra** [comment]
@KiChjang I got distracted by other projects and haven't finished the tests yet

**Pull request closed without merging**